packages/op-datasets/src/op_datasets/logic/transforms/daily_address_summary.py

```diff
@@ -0,0 +1,136 @@
+        ,SUM( CAST(receipt_l1_fee AS DECIMAL(38,18) ) / CAST( 1000000000000000000 AS DECIMAL(18, 0)) ) AS l1_contrib_gas_fees
+
+        ,SUM( CAST( (receipt_l1_gas_used * receipt_l1_blob_base_fee_scalar * receipt_l1_blob_base_fee) AS DECIMAL(38,18) )
+            / CAST( 1000000000000000000 AS DECIMAL(18, 0) )
+            ) AS l1_blobgas_contrib_gas_fees
+        ,SUM( CAST( (receipt_l1_gas_used * COALESCE(receipt_l1_base_fee_scalar,receipt_l1_fee_scalar) * receipt_l1_gas_price) AS DECIMAL(38,18) )
+            / CAST( 1000000000000000000 AS DECIMAL(18, 0) )
+            ) AS l1_l1gas_contrib_gas_fees
+
+        ,SUM(
+            CAST( (gas_price - max_priority_fee_per_gas) * receipt_gas_used AS DECIMAL(38,18) )
+            / CAST( 1000000000000000000 AS DECIMAL(18, 0) )
+        ) AS l2_contrib_gas_fees_base_fee
+        ,SUM(
+            CAST( max_priority_fee_per_gas * receipt_gas_used AS DECIMAL(38,18) )
+            / CAST( 1000000000000000000 AS DECIMAL(18, 0) )
+        ) AS l2_contrib_gas_fees_priority_fee
+
+        -- average fees
+        ,SUM( CAST( (gas_price * receipt_gas_used) AS DECIMAL(38,9) ) / CAST( 1000000000 AS DECIMAL(9, 0)) ) / SUM( CAST(receipt_gas_used AS DECIMAL(38,9)) ) AS avg_l2_gas_price_gwei
+        ,SUM( CAST( ((gas_price - max_priority_fee_per_gas) * receipt_gas_used) AS DECIMAL(38,9) ) / CAST( 1000000000 AS DECIMAL(9, 0)) ) / SUM( CAST(receipt_gas_used AS DECIMAL(38,9)) ) AS avg_l2_base_fee_gwei
+        ,SUM( CAST( (max_priority_fee_per_gas * receipt_gas_used) AS DECIMAL(38,9) ) / CAST( 1000000000 AS DECIMAL(9, 0)) ) / SUM( CAST(receipt_gas_used AS DECIMAL(38,9)) ) AS avg_l2_priority_fee_gwei
+        ,SUM( CAST( (receipt_l1_gas_price * receipt_l1_gas_used) AS DECIMAL(38,9) ) / CAST( 1000000000 AS DECIMAL(9, 0)) ) / SUM( CAST(receipt_l1_gas_used AS DECIMAL(38,9)) ) AS avg_l1_gas_price_gwei
+        ,SUM( CAST( (receipt_l1_blob_base_fee * receipt_l1_gas_used) AS DECIMAL(38,9) ) / CAST( 1000000000 AS DECIMAL(9, 0)) ) / SUM( CAST(receipt_l1_gas_used AS DECIMAL(38,9)) ) AS avg_l1_blob_base_fee_gwei
+
+        -- block time
+        ,MIN(block_timestamp) AS min_block_timestamp
+        ,MAX(block_timestamp) AS max_block_timestamp
+        ,MAX(block_timestamp) - MIN(block_timestamp) AS time_interval_active
+        ,COUNT(DISTINCT hour) AS unique_hours_active
+        ,CAST(COUNT(DISTINCT hour) AS FLOAT) / 24 AS perc_daily_hours_active
+
+        -- to addresses, to identify contracts in the future
+        ,COUNT(DISTINCT to_address) AS num_to_addresses
+        ,COUNT(DISTINCT CASE WHEN receipt_status = 1 THEN to_address ELSE NULL END) AS num_to_addresses_success
+        ,COUNT(DISTINCT CASE WHEN receipt_status != 1 THEN to_address ELSE NULL END) AS num_to_addresses_fail
+
+    FROM
+        _filter_df
+    GROUP BY
+        {groupby_clause}
+    """
+
+    return pl.sql(query).collect()
```

Comment:
One idea I had that can help people reading the model is to always have an assertion at the end that tells us what the schema of the result is.  That way one can get an idea of the structure that is being produced (and most importantly the exact dtypes being produced) without having to go in detail over the sql code. 

Could you give that a shot? It would look something like this:

```python
df = pl.sql(query).collect()
assert df.collect_schema() == {
  <column>: pl.<dtype>
  ...
}

return df
```

Reply:
Nice job @chuxinh ! This is a good start regarding an address summary, so I'll focus more on ergonomics and add extra features here in a separate PR.

### Some potential improvements were not possible with Polars:
- Polars does not support literals so you can't replace 1000000000 and 1000000000000000000 with 1e9 and 1e18 respectively which would improve readability a lot
- It also throws `SQLInterfaceError: multiple tables in FROM clause are not currently supported (found 2); use explicit JOIN syntax instead
` which rules out some options

### Materialized queries: 
- It would be nice if we were able to access or log the materialized query (query that actually runs). 
- Why? It would be useful to log the materialized query when we build more complex queries that have more variables down the line. 
- A naive solution would be to introduce `return query, pl.slq(query).collect()` , happy to hear your thoughts @lithium323 @chuxinh 

### Improving the query for readability:
I spent some time to refactor the query with the goal of making it more readable
- Constants: Defined two constants, WEI_IN_ETH and WEI_IN_GWEI, to avoid repeating the CAST operations throughout the query.
- tx_fees: Calculated all the fee-related expressions once, so they don't have to be recalculated multiple times in the main SELECT statement.

```
WITH constants AS (
    SELECT
        CAST(1000000000000000000 AS DECIMAL(18, 0)) AS WEI_IN_ETH,
        CAST(1000000000 AS DECIMAL(9, 0)) AS WEI_IN_GWEI
),
tx_fees AS (
    SELECT
        _filter_df.*,
        CAST(((gas_price * receipt_gas_used) + receipt_l1_fee) AS DECIMAL(38,18)) / constants.WEI_IN_ETH AS total_gas_fee,
        CASE WHEN receipt_status = 1 THEN
            CAST(((gas_price * receipt_gas_used) + receipt_l1_fee) AS DECIMAL(38,18)) / constants.WEI_IN_ETH
        ELSE 0 END AS total_gas_fee_success,
        CASE WHEN receipt_status != 1 THEN
            CAST(((gas_price * receipt_gas_used) + receipt_l1_fee) AS DECIMAL(38,18)) / constants.WEI_IN_ETH
        ELSE 0 END AS total_gas_fee_fail,
        CAST((gas_price * receipt_gas_used) AS DECIMAL(38,18)) / constants.WEI_IN_ETH AS l2_contrib_gas_fee,
        CAST(receipt_l1_fee AS DECIMAL(38,18)) / constants.WEI_IN_ETH AS l1_contrib_gas_fee,
        CAST((receipt_l1_gas_used * receipt_l1_blob_base_fee_scalar * receipt_l1_blob_base_fee) AS DECIMAL(38,18)) / constants.WEI_IN_ETH AS l1_blobgas_contrib_gas_fee,
        CAST((receipt_l1_gas_used * COALESCE(receipt_l1_base_fee_scalar, receipt_l1_fee_scalar) * receipt_l1_gas_price) AS DECIMAL(38,18)) / constants.WEI_IN_ETH AS l1_l1gas_contrib_gas_fee,
        CAST((gas_price - max_priority_fee_per_gas) * receipt_gas_used AS DECIMAL(38,18)) / constants.WEI_IN_ETH AS l2_contrib_gas_fee_base_fee,
        CAST(max_priority_fee_per_gas * receipt_gas_used AS DECIMAL(38,18)) / constants.WEI_IN_ETH AS l2_contrib_gas_fee_priority_fee,
        CAST((gas_price * receipt_gas_used) AS DECIMAL(38,9)) / constants.WEI_IN_GWEI AS total_gas_price_gwei,
        CAST(((gas_price - max_priority_fee_per_gas) * receipt_gas_used) AS DECIMAL(38,9)) / constants.WEI_IN_GWEI AS base_fee_gwei,
        CAST(max_priority_fee_per_gas * receipt_gas_used AS DECIMAL(38,9)) / constants.WEI_IN_GWEI AS priority_fee_gwei,
        CAST(receipt_l1_gas_price * receipt_l1_gas_used AS DECIMAL(38,9)) / constants.WEI_IN_GWEI AS l1_gas_price_gwei,
        CAST(receipt_l1_blob_base_fee * receipt_l1_gas_used AS DECIMAL(38,9)) / constants.WEI_IN_GWEI AS l1_blob_base_fee_gwei
    FROM
        _filter_df
    CROSS JOIN constants
)
SELECT
    {groupby_clause},

    -- Transactions
    COUNT(hash) AS total_txs,
    COUNT(CASE WHEN receipt_status = 1 THEN hash ELSE NULL END) AS total_txs_success,
    COUNT(CASE WHEN receipt_status != 1 THEN hash ELSE NULL END) AS total_txs_fail,

    -- Blocks
    COUNT(DISTINCT block_number) AS total_blocks,
    COUNT(DISTINCT CASE WHEN receipt_status = 1 THEN block_number ELSE NULL END) AS total_blocks_success,
    COUNT(DISTINCT CASE WHEN receipt_status != 1 THEN block_number ELSE NULL END) AS total_blocks_fail,
    MIN(block_number) AS min_block_number,
    MAX(block_number) AS max_block_number,
    MAX(block_number) - MIN(block_number) + 1 AS block_interval_active,

    -- Nonce
    MIN(nonce) AS min_nonce,
    MAX(nonce) AS max_nonce,
    MAX(nonce) - MIN(nonce) + 1 AS nonce_interval_active,

    -- Gas Usage
    SUM(receipt_gas_used) AS total_l2_gas_used,
    SUM(CASE WHEN receipt_status = 1 THEN receipt_gas_used ELSE 0 END) AS total_l2_gas_used_success,
    SUM(CASE WHEN receipt_status != 1 THEN receipt_gas_used ELSE 0 END) AS total_l2_gas_used_fail,
    SUM(receipt_l1_gas_used) AS total_l1_gas_used,
    SUM(CASE WHEN receipt_status = 1 THEN receipt_l1_gas_used ELSE 0 END) AS total_l1_gas_used_success,
    SUM(CASE WHEN receipt_status != 1 THEN receipt_l1_gas_used ELSE 0 END) AS total_l1_gas_used_fail,

    -- Gas Fee Paid
    SUM(total_gas_fee) AS total_gas_fees,
    SUM(total_gas_fee_success) AS total_gas_fees_success,
    SUM(total_gas_fee_fail) AS total_gas_fees_fail,

    -- Fee Breakdown
    SUM(l2_contrib_gas_fee) AS l2_contrib_gas_fees,
    SUM(l1_contrib_gas_fee) AS l1_contrib_gas_fees,
    SUM(l1_blobgas_contrib_gas_fee) AS l1_blobgas_contrib_gas_fees,
    SUM(l1_l1gas_contrib_gas_fee) AS l1_l1gas_contrib_gas_fees,
    SUM(l2_contrib_gas_fee_base_fee) AS l2_contrib_gas_fees_base_fee,
    SUM(l2_contrib_gas_fee_priority_fee) AS l2_contrib_gas_fees_priority_fee,

    -- Average Fees
    SUM(total_gas_price_gwei) / SUM(CAST(receipt_gas_used AS DECIMAL(38,9))) AS avg_l2_gas_price_gwei,
    SUM(base_fee_gwei) / SUM(CAST(receipt_gas_used AS DECIMAL(38,9))) AS avg_l2_base_fee_gwei,
    SUM(priority_fee_gwei) / SUM(CAST(receipt_gas_used AS DECIMAL(38,9))) AS avg_l2_priority_fee_gwei,
    SUM(l1_gas_price_gwei) / SUM(CAST(receipt_l1_gas_used AS DECIMAL(38,9))) AS avg_l1_gas_price_gwei,
    SUM(l1_blob_base_fee_gwei) / SUM(CAST(receipt_l1_gas_used AS DECIMAL(38,9))) AS avg_l1_blob_base_fee_gwei,

    -- Block Time
    MIN(block_timestamp) AS min_block_timestamp,
    MAX(block_timestamp) AS max_block_timestamp,
    MAX(block_timestamp) - MIN(block_timestamp) AS time_interval_active,
    COUNT(DISTINCT hour) AS unique_hours_active,
    CAST(COUNT(DISTINCT hour) AS FLOAT) / 24 AS perc_daily_hours_active,

    -- To Addresses
    COUNT(DISTINCT to_address) AS num_to_addresses,
    COUNT(DISTINCT CASE WHEN receipt_status = 1 THEN to_address ELSE NULL END) AS num_to_addresses_success,
    COUNT(DISTINCT CASE WHEN receipt_status != 1 THEN to_address ELSE NULL END) AS num_to_addresses_fail

FROM
    tx_fees
GROUP BY
    {groupby_clause};
```

## Comparing results to make sure we have the same data:

Spot checks:
<img width="1375" alt="image" src="https://github.com/user-attachments/assets/32dbd7b0-a900-4578-9e8e-f53739ecd9be">

When comparing both dataframes we only get rounding errors in 3 columns

<img width="1232" alt="image" src="https://github.com/user-attachments/assets/543f4a68-bcb6-4c6a-ae1e-fdc3c5db1f90">

Verified that the max difference between the values is les than 1e-17
<img width="731" alt="image" src="https://github.com/user-attachments/assets/72af2c83-3b0c-4b2a-b362-e072418a05dd">